### PR TITLE
basic/memfd: untangle historic `sealed` property

### PR DESCRIPTION
**Pull request by dvdhrm**
A series of minor changes to `basic/memfd` which untangles the historic `is/set_sealed()` functionality. With the upstream memfd seals, there is no need to always set and require the full set of seals. Instead, ensure those functions work like the historic out-of-tree sealing again.

See each commit-message for details.

### CODE REVIEW

File: src/basic/memfd-util.c [hunk 1]
```diff
@@ -68,6 +68,26 @@ int memfd_new(const char *name) {
         return memfd_create_wrapper(name, MFD_ALLOW_SEALING | MFD_CLOEXEC | MFD_NOEXEC_SEAL);
 }
 
+int memfd_add_seals(int fd, unsigned int seals) {
+        assert(fd >= 0);
+
+        return RET_NERRNO(fcntl(fd, F_ADD_SEALS, seals));
+}
+
+int memfd_get_seals(int fd, unsigned int *sealsp) {
```
Comment (bluca) [inline src/basic/memfd-util.c]: ret_seals

**Review by bluca — CHANGES_REQUESTED** [review]
> This was recently relaxed, since
newer memfds have `F_SEAL_EXEC` set by default. This broke live-updates
of systemd. Similarly, this patch will break live-udpates, by reducing
the set of seals and thus relying on the receive end to only test for
the reduced set. Given that this worked fine for `F_SEAL_EXEC`, it seems
ok to do this now again.

I'm not sure this is actually ok. But why do it in the first place? What's the goal with removing F_SEAL_SEAL?

**Comment by dvdhrm** [comment]
> > This was recently relaxed, since
> > newer memfds have `F_SEAL_EXEC` set by default. This broke live-updates
> > of systemd. Similarly, this patch will break live-udpates, by reducing
> > the set of seals and thus relying on the receive end to only test for
> > the reduced set. Given that this worked fine for `F_SEAL_EXEC`, it seems
> > ok to do this now again.
> 
> I'm not sure this is actually ok. But why do it in the first place? What's the goal with removing F_SEAL_SEAL?

There is no reason to test the flag or set it, other than keeping historic behavior. Hence, I thought we might just drop it.

I can include it in `memfd_set_sealed()`, if you prefer.

**Comment by bluca** [comment]
> > > This was recently relaxed, since
> > > newer memfds have `F_SEAL_EXEC` set by default. This broke live-updates
> > > of systemd. Similarly, this patch will break live-udpates, by reducing
> > > the set of seals and thus relying on the receive end to only test for
> > > the reduced set. Given that this worked fine for `F_SEAL_EXEC`, it seems
> > > ok to do this now again.
> > 
> > 
> > I'm not sure this is actually ok. But why do it in the first place? What's the goal with removing F_SEAL_SEAL?
> 
> There is no reason to test the flag or set it, other than keeping historic behavior. Hence, I thought we might just drop it.
> 
> I can include it in `memfd_set_sealed()`, if you prefer.

If updates can break and there's not much to gain to remove it, then yes please let's keep it for now

**Comment by dvdhrm** [comment]
(v2: use `ret_seals` and retain `F_SEAL_SEAL` for outgoing memfds)

**Review by bluca — APPROVED** [review]

**Pull request merged**